thunks.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2017 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonnet
+
+// readyValue
+// -------------------------------------
+
+// readyValue is a wrapper which allows to use a concrete value where normally
+// some evaluation would be expected (e.g. object fields). It's not part
+// of the value interface for increased type safety (it would be very easy
+// to "overevaluate" otherwise) and conveniently it also saves us from implementing
+// these methods for all value types.
+type readyValue struct {
+	content value
+}
+
+func (rv *readyValue) getValue(i *interpreter, t *TraceElement) (value, error) {
+	return rv.content, nil
+}
+
+func (rv *readyValue) bindToObject(sb selfBinding, origBinding bindingFrame) potentialValue {
+	return rv
+}
+
+// potentialValues
+// -------------------------------------
+
+// thunk holds code and environment in which the code is supposed to be evaluated
+type thunk struct {
+	name identifier
+	env  environment
+	body astNode
+}
+
+func makeThunk(name identifier, env environment, body astNode) *cachedThunk {
```

Comment:
makeThunk returning a cachedThunk is weird.  Maybe call thunk 'exprThunk' (or astThunk but then it looks like an AST node).  Then call cachedThunk just thunk?  Or, call this makeCachedExprThunk because that's what it really is.

Reply:
I like makeCachedExprThunk the most. But I want to think about it a bit more and I have another idea that I may want to try. So I've added a TODO with your comment, I will look into it soon.

Reply:
btw I've also changed the return type everywhere to potentialValue to express the intention better